### PR TITLE
Undo and toast

### PR DESCRIPTION
**Pull request by bluquar**
https://user-images.githubusercontent.com/4309735/149575548-6009977e-839a-41e5-9fb2-3196f383f4f4.mov

https://user-images.githubusercontent.com/4309735/149575553-290187ca-7cf1-4ab4-b018-c491aa780c6a.mov

### CODE REVIEW

File: src/client/App.tsx [hunk 1]
```diff
@@ -36,10 +39,12 @@ export default function App() {
...
-            </DrawerProvider>
+            <ToastProvider>
+              <DrawerProvider>
+                <Navigation colorScheme={colorScheme} />
+                <StatusBar />
+              </DrawerProvider>
+            </ToastProvider>
           </ApolloProvider>
         </PaperProvider>
       </SafeAreaProvider>
```
Comment (lowellbander) [inline src/client/App.tsx:50]: hehe i think this provider sandwich is funny

File: src/client/aid-app/aid_requests/CreateRequestForm.tsx [hunk 2]
```diff
@@ -54,7 +50,7 @@ export default function CreateRequestForm({
     </>
   );
   async function submit(): Promise<void> {
-    setShowCreationToast(false);
+    publishToast(undefined);
```
Comment (lowellbander) [inline src/client/aid-app/aid_requests/CreateRequestForm.tsx:53]: someone may wanna undo form submission, right?

File: src/client/general-purpose/toast/ToastContext.ts [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+  undo?: null | (() => Promise<void>);
+};
+
+export type ToastContextType = {
+  publishToast: (config: ToastConfig | undefined) => void;
+};
+
+const ToastContext = React.createContext<ToastContextType>({
+  publishToast: () => {
+    console.error('Attempted to publish toast outside of a ToastContext');
```
Comment (lowellbander) [inline src/client/general-purpose/toast/ToastContext.ts:14]: contexts/providers seem very peculiar to me: on the one hand they require this dynamic error handling, and on the other the Providers are declared as JSX components which seems more odd than some other config option

just musinggg

File: src/client/general-purpose/toast/ToastProvider.tsx [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+      publishToast: setToastContext,
+    }),
+    [],
+  );
+
+  return (
+    <ToastContext.Provider value={value}>
+      {children}
+      <View style={[styles.snackbar, { minWidth: windowWidth - 32 }]}>
+        <Snackbar
```
Comment (lowellbander) [inline src/client/general-purpose/toast/ToastProvider.tsx:27]: nom nom nom

File: src/server/collections/aid_request/helpers/getWhoRecordedRequest.ts [hunk 5]
```diff
@@ -0,0 +1,13 @@
+import { UserModel } from '../../user/UserModel';
+import type { AidRequestType } from '../AidRequestModelTypes';
+
+export default async function getWhoRecordedRequest(
```
Comment (lowellbander) [inline src/server/collections/aid_request/helpers/getWhoRecordedRequest.ts:4]: was this just a move? I made this, right? or was it sth else?

File: src/server/collections/aid_request/object_fields/latestEvent.ts [hunk 6]
```diff
@@ -63,3 +66,36 @@ function getActionText(details: AidRequestHistoryEventPayload): string {
 }
 
 export default latestEvent;
+
+function filterRemovals(
+  history: Array<AidRequestHistoryEvent>,
+): Array<AidRequestHistoryEvent> {
+  const toRemove = new Set<AidRequestHistoryEvent>();
+  const unpairedAdds: Map<string, AidRequestHistoryEvent> = new Map();
```
Comment (lowellbander) [inline src/server/collections/aid_request/object_fields/latestEvent.ts:74]: hehe feels like an interview question

**Review by lowellbander — APPROVED** [review]
fuckin shit real cool!

**Comment by lowellbander** [comment]
"6 conversations must be resolved before merging."
"must"

File: src/client/aid-app/aid_requests/CreateRequestForm.tsx (hunk 2 above)
Comment (bluquar) [reply src/client/aid-app/aid_requests/CreateRequestForm.tsx:53]: Yes. Not yet implemented :)

File: src/server/collections/aid_request/helpers/getWhoRecordedRequest.ts (hunk 5 above)
Comment (bluquar) [reply src/server/collections/aid_request/helpers/getWhoRecordedRequest.ts:4]: Yep. It's a move with some changes

File: src/server/collections/aid_request/object_fields/latestEvent.ts (hunk 6 above)
Comment (bluquar) [reply src/server/collections/aid_request/object_fields/latestEvent.ts:74]: Feels like the kind of thing that should have unit tests lol

**Pull request merged**